exporter/elasticsearchexporter/integrationtests/exporter_test.go

```diff
@@ -0,0 +1,161 @@
+		{name: "collector_restart_with_es_intermittent_failure", mockESFailure: true, restartCollector: true},
+		*/
+	} {
+		t.Run(tc.name, func(t *testing.T) {
+			runner(t, tc.restartCollector, tc.mockESFailure)
+		})
+	}
+}
+
+func runner(t *testing.T, restartCollector, mockESFailure bool) {
+	t.Helper()
+
+	cfg, err := loadConfig()
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	mockES, err := newMockESClient(t, cfg.Debug)
+	require.NoError(t, err)
+
+	collector, err := newTestCollector(t, cfg, mockES.ServerURL)
+	require.NoError(t, err)
+	t.Cleanup(collector.Shutdown)
+
+	var g errgroup.Group
+	cancelRun := runTestCollectorWithWait(ctx, t, collector, &g)
+
+	if mockESFailure {
+		mockES.SetReturnStatusCode(http.StatusServiceUnavailable)
+		if !restartCollector {
+			// If restartCollector is not set then recover ES service
+			// after a few failed bulk requests otherwise ensure that
+			// both collector and ES are unavailable for the same
+			// duration considering worst case scenario.
+			time.AfterFunc(10*time.Second, func() {
+				mockES.SetReturnStatusCode(http.StatusOK)
+			})
+		}
+	}
+
+	var totalLogCount int
+	sendLogs(t, cfg.GRPCEndpoint, "batch_1", 2_000 /* log count */, 10)
+	totalLogCount += 2_000
+
+	if restartCollector {
+		// Restart the collector after all data is sent to the collector.
+		// Note that the log count should be high to ensure that the collector
+		// stops before all data is shipped to Elasticsearch.
+		require.NoError(t, collector.Recreate())
+		cancelRun()
+		// Ensure that the previous collector instance is shut down and then
+		// run the recreated collector.
+		require.NoError(t, g.Wait())
+		runTestCollectorWithWait(ctx, t, collector, &g)
+		// Recover ES if it was unavailable.
+		mockES.SetReturnStatusCode(http.StatusOK)
+	}
+
+	sendLogs(t, cfg.GRPCEndpoint, "batch_2", 2_000 /* log count */, 10)
+	totalLogCount += 2_000
+
+	assert.Eventually(
+		t, func() bool {
+			resp, err := mockES.Count(mockES.Count.WithIndex(cfg.ESLogsIndex))
+			require.NoError(t, err)
+
+			body, err := ioutil.ReadAll(resp.Body)
+			require.NoError(t, err)
+
+			result := gjson.GetBytes(body, "count")
+			return result.Int() == int64(totalLogCount)
+		}, time.Minute, time.Second,
+	)
+}
+
+// runTestCollectorWithWait runs the test OTEL collector in a goroutine.
+// It returns a cancel func that allows the caller to stop the collector
+// without waiting for graceful shutdown.
+func runTestCollectorWithWait(
+	ctx context.Context, t *testing.T, collector *otelCol, g *errgroup.Group,
+) context.CancelFunc {
+	t.Helper()
+
+	rCtx, rCancel := context.WithCancel(ctx)
+	g.Go(func() error { return collector.Run(rCtx) })
+	// Wait for otelcollector to be in running state
+	require.Eventually(t, func() bool {
+		return collector.IsRunning()
+	}, time.Second, 10*time.Millisecond)
+	return rCancel
+}
+
+func sendLogs(t *testing.T, target, uid string, logs, agents int) {
```

Comment:
[**For Reviewers**] I wanted to use [telemetrygen](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/cmd/telemetrygen) here instead of writing something new but it is only exposed as a command. Does it make sense to refactor `telemetrygen` as a follow-up and make it a bit easier to use as a package?